man/getmap.3
.TH GETMAP 3 "12 September 1995"
.FB
.SH NAME
getmap, putmap \- read and write database map entry
.SH SYNOPSIS
#include <firstbase/fb.h>
.sp 1
fb_getmap(mfd, n, avail, freep, rpos, rlen)
.br
int mfd;
.br
long n, *avail, *freep, *rpos, *rlen;
.br
.sp 1
fb_putmap(mfd, n, avail, rpos, rlen)
.br
int mfd;
.br
long n, avail, rpos, rlen;
.PP
.SH DESCRIPTION
As explained in the
\fIdatabase\fP(4) manual page, a database structure contains
all of the elements needed for an entire database, including
file names, descriptors and other variables. One of the files associated
with a FirstBase database is a map file used to provide locations and
lengths of logical FirstBase records.
.PP
This database map file, actually a miniature, fixed-width database in its
own right, is modified using the low level FirstBase routines \fIfb_getmap\fP
and \fIfb_putmap\fP. However, these routines are only used when building a
\fIFirstBase\fP map file without using \fIputrec\fP(3).
.PP
Formally, \fIfb_getmap\fP reads a map "record" plus header information
from the map file, and \fIfb_putmap\fP writes a map "record" to the map file.
.PP
The \fIfb_getmap\fP and \fIfb_putmap\fP routines, being very low level,
do not know of database structures and such.
Instead, each is passed all the arguments
it needs to retrieve a map entry. For \fIfb_getmap\fP the arguments are:
\fImfd\fP, the map file descriptor,
as in db->mfd; \fIn\fP is the logical record number for the map, and in
turn, the information returned is about the database record known as record
\fIn\fP; \fIavail\fP, \fIfreep\fP, \fIrpos\fP, and \fIrlen\fP are all
long pointers (long *) and are used as storage for the \fIfb_getmap\fP request.
Actually, \fIrpos\fP and \fIrlen\fP are stored only if \fIn\fP is non-zero.
.PP
The information returned from \fIfb_getmap\fP is as follows:
\fIavail\fP is the byte location of the next free byte in the database file;
\fIfreep\fP is the byte location of the head of the free list, inside
the database file; \fIrpos\fP is the record position (byte location)
and \fIrlen\fP is the record length (in bytes).
.PP
Map records and/or the map header
can be updated or written to the database map file using the
\fIfb_putmap\fP mechanism. \fIPutmap\fP takes most of the same arguments
as \fIfb_getmap\fP: \fImfd\fP and \fIn\fP are exactly the same.
However, there is not a \fIfreep\fP value (the free list is maintained
by separate FirstBase low level routines). The other three arguments
represent the same variables: \fIavail\fP, \fIrpos\fP,
and \fIrlen\fP. Note that \fIrpos\fP and \fIrlen\fP are written only
when referencing a non-zero record. Record number zero (0) is the map header.
.PP
NOTE: \fIfb_putmap\fP and \fIfb_getmap\fP are usually not needed unless you
are generating FirstBase databases using buffered writes (see
$FIRSTBASEHOME/applications/sample_src/bload.c) or something similar. 
.SH SEE ALSO
database(4), opendb(3).
.SH DIAGNOSTICS
If \fIfb_putmap\fP or \fIfb_getmap\fP fail due to a read or seek error,
they will return FB_ERROR. If all goes correctly, an FB_AOK status is returned.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/librec/getmap.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: getmap.c,v 9.0 2001/01/09 02:56:58 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Getmap_sid[] = "@(#) $Id: getmap.c,v 9.0 2001/01/09 02:56:58 john Exp $";
#endif

#include <fb.h>
#include <fb_ext.h>

extern short int cdb_dbase_byteorder;
extern short int cdb_cpu_byteorder;

/* 
 *  getmap - get map information - avail, freep. also rpos,rlen if n > 0.
 */
 
   fb_getmap(mfd, n, avail, freep, rpos, rlen)
      long *avail, *freep, *rpos, *rlen;
      long n;
      int mfd;
   
      {
         long fend, ns;
         char buf[FB_SLONG+FB_SLONG+3];
         int map_rlen = FB_SLONG + FB_SLONG;
	 
	 if (lseek(mfd, 0L, 0) < 0L){
	    fb_lerror(FB_SEEK_ERROR, SYSMSG[S_BAD_MAP], NIL);
	    return(FB_ERROR);
	    }
	 if (read(mfd, buf, (unsigned) map_rlen) != map_rlen){
	    fb_lerror(FB_READ_ERROR, SYSMSG[S_BAD_MAP], NIL);
	    return(FB_ERROR);
	    }
         memcpy((char *) avail, buf, FB_SLONG);
         memcpy((char *) freep, buf + FB_SLONG, FB_SLONG);
         if (cdb_dbase_byteorder != cdb_cpu_byteorder){
            M_32_SWAP((*avail));
            M_32_SWAP((*freep));
            }
	 if (n <= 0L)
	    return(FB_AOK);
	 if ((fend = lseek(mfd, 0L, 2)) < 0L){
	    fb_lerror(FB_SEEK_ERROR, SYSMSG[S_BAD_MAP], NIL);
	    return(FB_ERROR);
	    }
	 ns = n * (long) FB_MAPREC_SIZE;
	 if (ns < fend){
	    if (lseek(mfd, ns, 0) < 0L){
	       fb_lerror(FB_SEEK_ERROR, SYSMSG[S_BAD_MAP], NIL);
	       return(FB_ERROR);
	       }
            if (read(mfd, buf, (unsigned) map_rlen) != map_rlen){
               fb_lerror(FB_READ_ERROR, SYSMSG[S_BAD_MAP], NIL);
               return(FB_ERROR);
               }
            memcpy((char *) rpos, buf, FB_SLONG);
            memcpy((char *) rlen, buf + FB_SLONG, FB_SLONG);
            if (cdb_dbase_byteorder != cdb_cpu_byteorder){
               M_32_SWAP((*rpos));
               M_32_SWAP((*rlen));
               }
	    }
	 else{
	    *rpos = 0L;
	    *rlen = 0L;
	    }
	 return(FB_AOK);
      }
